DATASET_VERSIONING.md

```diff
@@ -0,0 +1,76 @@
+# Dataset versioning (DRAFT)
+
+DSO API provides option to serve multiple versions of same dataset via API and Database.
+
+This document aims to cover technical details on how DSO API behaves in different scenarios, 
+depending on Dataset model configuration.
+
+
+## How dataset version is defined
+
+`Dataset` model exposes 2 parameters to control versioning:
+
+ - `Dataset.version`: string, formatted according to [Semantic Versioning standard](https://semver.org/) e.g. `0.1.0` 
+   (`<major> "." <minor> "." <patch>`).
+ - `Dataset.is_default_version`: boolean, defaults to `True`
+
+Combination of these 2 parameters defines how API and Database behave.
+
+
+## Datasets without version definition and default versions of datasets
+
+All datasets without version definition in `Dataset.version` or with `Dataset.is_default_version == True` 
+will be treated as not versioned and all tables within dataset will have API endpoint set to 
+
+> `/v1/<dataset.url_prefix>/<dataset.id>/<table.name>/`
+
+Database tables of this dataset will be prefixed with `<dataset.id>_`
+
+
+## Datasets with non-default versions
+
+Any dataset wit `version` defined and `is_default_version` is set to `False` will have API endpoint defined as:
+
+ > `/v1/<dataset.url_prefix>/<dataset.id>@<dataset.version>/<table.name>/`
+
+
+## Database structure
+
+All tables within default version of Dataset will have `<dataset.id>_` name prefix.
+
+Other Dataset versions will have `<dataset.id>_<dataset.version.major>_<dataset.version.minor>_` table name prefix.
+
+> DRAFT NOTE:
+> 
+>`<dataset.version.major>_<dataset.version.minor>` should be used with caution 
+as underscore can be lost in `to_snake_case` -> `toCamelCase` -> `to_snake_case` translation, 
+however should not come into play as table name is rather isolated.
```

Comment:
Why not circumvent the whole `to_snake_case` -> `toCamelCase` -> `to_snake_case` problem by using `$` as the separator between major and minor version number? PostgreSQL [allows for it](https://www.postgresql.org/docs/12/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS):

> SQL identifiers and key words must begin with a letter (a-z, but also letters with diacritical marks and non-Latin letters) or an underscore (_). Subsequent characters in an identifier or key word can be letters, underscores, digits (0-9), or dollar signs ($). Note that dollar signs are not allowed in identifiers according to the letter of the SQL standard, so their use might render applications less portable. The SQL standard will not define a key word that contains digits or starts or ends with an underscore, so identifiers of this form are safe against possible conflict with future extensions of the standard.

It's ugly to use `$` but anything better than having major and minor version number suddenly being lumped together. (eg `1_3` suddenly becoming `13`)

Reply:
Dollar sign will do as well, however it might bring more issues in future. 

We do not expect database table name conversions, so underscore should safe to use and less ugly than dollar sign.

Reply:
> We do not expect database table name conversions

But if the need does arrive, we've got a big problem. We can plan for that, though, by putting the version last: `test_users_0_2` instead of `test_0_2_users` (or even `test_users_v0_2`?). That way, it'll be easier to parse the version number off the table name before attempting any conversion.

Reply:
Using version suffix might be better approach. 
Though it has disadvantage: it will be harder to search for tables that belong to same version and that might be bigger issue with cross-dataset relation tables, where version of both datasets should be taken into account.

Reply:
You mean because `users_.*_1_0` matches `users_id_1_1__dogs_id_1_0`?